tests/test_upgrade_shared.py

```diff
@@ -0,0 +1,62 @@
+import subprocess
+
+from helpers import run_pipx_cli
+
+
+def test_upgrade_shared(pipx_ultra_temp_env, capsys, caplog, monkeypatch):
+    from pipx.shared_libs import shared_libs
+
+    assert shared_libs.has_been_updated_this_run is False
+    assert shared_libs.is_valid is False
+    assert run_pipx_cli(["upgrade-shared", "-v"]) == 0
+    captured = capsys.readouterr()
+    assert "creating shared libraries" in captured.err
+    assert "upgrading shared libraries" in captured.err
+    assert "Upgrading shared libraries in" in caplog.text
+    assert "Already upgraded libraries in" not in caplog.text
+    assert shared_libs.has_been_updated_this_run is True
+    assert shared_libs.is_valid is True
+    shared_libs.has_been_updated_this_run = False
+    assert run_pipx_cli(["upgrade-shared", "-v"]) == 0
+    captured = capsys.readouterr()
+    assert "creating shared libraries" not in captured.err
+    assert "upgrading shared libraries" in captured.err
+    assert "Upgrading shared libraries in" in caplog.text
+    assert "Already upgraded libraries in" not in caplog.text
+    assert shared_libs.has_been_updated_this_run is True
+    assert run_pipx_cli(["upgrade-shared", "-v"]) == 0
+    assert "Already upgraded libraries in" in caplog.text
+
+
+def test_upgrade_shared_pip_args(pipx_ultra_temp_env, capsys, caplog, monkeypatch):
+    from pipx.shared_libs import shared_libs
+
+    assert shared_libs.has_been_updated_this_run is False
+    assert shared_libs.is_valid is False
+    assert run_pipx_cli(["upgrade-shared", "-v", "--pip-args='--no-index'"]) == 1
+    captured = capsys.readouterr()
+    assert "creating shared libraries" in captured.err
+    assert "upgrading shared libraries" in captured.err
+    assert "Upgrading shared libraries in" in caplog.text
+    assert "Already upgraded libraries in" not in caplog.text
+    assert shared_libs.has_been_updated_this_run is False
+    assert shared_libs.is_valid is True
+
+
+def test_upgrade_shared_pin_pip(pipx_ultra_temp_env, capsys, caplog, monkeypatch):
+    from pipx.shared_libs import shared_libs
+
+    def pip_version():
```

Comment:
We might make a test fixture out of this?

Reply:
I don't see how but feel free to push on the branch.

Reply:
Ah, I see. This isn't feasible because the Python interpreter for the shared lib doesn't exist at the time the fixture is called. I thought of creating a general way to retrieve the `pip` version based on a Python interpreter path.